File: C++/프로그래머스/1/1845. 폰켓몬/폰켓몬.cpp
#include <vector>   
#include <algorithm> 
#include <unordered_set> 

using namespace std;

// 폰켓몬 종류를 선택하는 최대 종류 수를 반환하는 함수
int solution(vector<int> nums) {
    // 1. 중복을 제거하기 위해 unordered_set을 사용하여 고유한 폰켓몬 종류를 저장
    unordered_set<int> unique_poketmons(nums.begin(), nums.end());

    // 2. 선택할 수 있는 폰켓몬의 수는 전체 폰켓몬 수의 절반
    int max_select = nums.size() / 2;

    // 3. 최대 종류 수는 선택 가능한 수와 고유한 폰켓몬 수 중 작은 값
    return min(max_select, (int)unique_poketmons.size());
}

File: Java/프로그래머스/1/1845. 폰켓몬/폰켓몬.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int solution(int[] nums) {                
        int n = nums.length / 2;  // 배열의 크기의 절반을 계산
       
        // 배열에서 중복을 제거한 Set을 생성
        Set<Integer> num_set = new HashSet<>();
        for (int num : nums) {
            num_set.add(num);
        }
        
        // Set의 크기와 n 중 작은 값을 선택
        int answer = Math.min(n, num_set.size());
        
        return answer;
    }
}
